### PR TITLE
Feat/add token

### PR DESCRIPTION
**Pull request by Butcher-L**
fixes PR and refactor

### CODE REVIEW

File: source/controllers/users.ts [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
-import SignJwt from '../functions/sign-jwt';
 import Config from '../configs/config'
 import Jwt from 'jsonwebtoken'
 
-
-
+import Auth from 'basic-auth'
 
 const addUser = async(ctx ,next) => {
     let { username, password } = ctx.request.body;
```
Comment (ccpacillos) [inline source/controllers/users.ts]: `username` and `password` pair should be from the basic auth header. In the specs from the exercise, only `name` is in the request body.

File: source/controllers/users.ts [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
 
+
     const user = new Users({
+        name,
         username,
         password: await Bcryptjs.hash(password, 10)
     })
```
Comment (ccpacillos) [inline source/controllers/users.ts]: Rename `find` to `userExists`.

File: source/controllers/users.ts [hunk 3]
```diff
@@ -24,31 +22,34 @@ const addUser = async(ctx ,next) => {
         ctx.body='User already exist'
     }else{
         const res = await user.save()
```
Comment (ccpacillos) [inline source/controllers/users.ts]: Just use `User.create({ ... });` here.

File: source/controllers/users.ts [hunk 4]
```diff
@@ -24,31 +22,34 @@ const addUser = async(ctx ,next) => {
...
 }
 
 const login = async(ctx ,next) => {
-    let {username, password} = ctx.request.body;
+    
+    const authCredentials = Auth(ctx)
+    const username = authCredentials.name
+    const password = authCredentials.pass
 
     try {
```
Comment (ccpacillos) [inline source/controllers/users.ts]: Remove unnecessary `try-catch`.

File: source/controllers/users.ts [hunk 5]
```diff
@@ -24,31 +22,34 @@ const addUser = async(ctx ,next) => {
...
-
         if(!user){
             ctx.status = 400
-            throw new Error("Not found")
+            ctx.message='User not registerd'
+            return
         }
 
-        const Bcrypt = await Bcryptjs.compare(password, user.password)
+        const Bcrypt = await Bcryptjs.compare(password,user.password)
```
Comment (ccpacillos) [inline source/controllers/users.ts]: Use a more sensible name, that represents a `boolean` value that indicates whether the password is correct or not.

File: source/middleware/sign-jwt.ts [hunk 6]
```diff
@@ -9,7 +9,8 @@ const  signJWT = async(user: IUser, callback: (error:Error | null, token:string
 
     try {
```
Comment (ccpacillos) [inline source/middleware/sign-jwt.ts]: Remove unnecessary try catch.

File: source/models/users.ts [hunk 7]
```diff
@@ -3,6 +3,7 @@ import IUsers from '../interfaces/users';
 
 const UserSchema: Schema = new Schema(
     {
+        name:{type: String, require: true },
         username: { type: String, require: true },
```
Comment (ccpacillos) [inline source/models/users.ts]: Set the `username` to be unique.

File: source/test/users/users.spec.ts [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
 afterEach(() => {
-    Users.deleteMany({name:randomUsername}).exec()
+    Users.deleteMany({username:randomUsername}).exec()
 })
 let randomUsername,
     randomPassword,
     presentUsername,
-    presentPassword
+    presentPassword,
+    token
```
Comment (ccpacillos) [inline source/test/users/users.spec.ts]: Use `const` within your `it()` test cases to ensure there's no sharing of values between test cases.

File: source/test/users/users.spec.ts [hunk 9]
```diff
@@ -19,42 +20,11 @@ describe("Users Test", ()=>{
         randomPassword = internet.password();
     })
 
-    it("Add new user - POST/users", async() =>{
-        presentUsername=randomUsername
-        presentPassword=randomPassword
-        await Request(Server).post("/users")
+    it("Login - POST/auth", async() =>{
+        const res = await Request(Server).post("/auth")
         .send({username:"Butch",password:"1234"})
```
Comment (ccpacillos) [inline source/test/users/users.spec.ts]: Don't use static values. You can make use of `chance` npm module to generate random values.

**Review by ccpacillos — CHANGES_REQUESTED** [review]

File: source/controllers/users.ts [hunk 10]
```diff
@@ -1,76 +1,73 @@
...
+    }
 
-        const Bcrypt = await Bcryptjs.compare(password, user.password)
-        if(Bcrypt){
-            const timeInMilliseconds = new Date().getTime();
-            const expirationTime  = timeInMilliseconds + Number(Config.token.expireTime) * 10_000;
-            const expireTimeInSeconds = Math.floor(expirationTime/1_000);
+    const ifMatch = await Bcryptjs.compare(password, user.password);
+    
+    if (ifMatch) {
```
Comment (ccpacillos) [inline source/controllers/users.ts]: Not a critical change. For this one, it's recommended that variables are easy to read.
`ifMatch` could be `passwordIsValid`. It comes very useful during discussions you will have with your teammates.

It's easier to say "If password is valid, then...", compared to "If if match, then".

File: source/interfaces/users.ts [hunk 11]
```diff
@@ -1,6 +1,7 @@
 import { Document } from 'mongoose';
 
 export default interface IUser extends Document{
```
Comment (ccpacillos) [inline source/interfaces/users.ts]: Use `User` rather than `IUser`.

File: source/middleware/sign-jwt.ts [hunk 12]
```diff
@@ -7,9 +7,9 @@ const  signJWT = async(user: IUser, callback: (error:Error | null, token:string
     const expirationTime  = timeInMilliseconds + Number(Config.token.expireTime) * 10_000;
```
Comment (ccpacillos) [inline source/middleware/sign-jwt.ts]: I don't think you have used this anywhere. So you can just remove it.

File: source/test/users/users.spec.ts [hunk 13]
```diff
@@ -1,60 +1,33 @@
...
-afterEach(() => {
-    Users.deleteMany({name:randomUsername}).exec()
-})
-let randomUsername,
-    randomPassword,
-    presentUsername,
-    presentPassword
-
-describe("Users Test", ()=>{
+describe.only("Users Test", ()=>{
```
Comment (ccpacillos) [inline source/test/users/users.spec.ts]: Remove `.only`.

**Review by ccpacillos — CHANGES_REQUESTED** [review]

File: source/controllers/users.ts (hunk 10 above)
Comment (Butcher-L) [reply source/controllers/users.ts]: copy sir

File: source/middleware/sign-jwt.ts (hunk 12 above)
Comment (Butcher-L) [reply source/middleware/sign-jwt.ts]: I use it to calculate the expiration in seconds

Comment (ccpacillos) [reply source/middleware/sign-jwt.ts]: Oh sorry, what I meant was the file `middleware/sign-jwt.ts` was not used anywhere.

**Review by ccpacillos — APPROVED** [review]

**Pull request merged**